docs/design-doc.md

```diff
@@ -0,0 +1,296 @@
+
+The Relay service runs as a standalone process that is separated by a network boundary. Inbound DWeb Messages are received by the relay service, integrity checked, and then relayed (aka proxied) to the downstream service.
+
+Integration between the relay service and the downstream service is achieved by registering (a.k.a configuring) a _Handler_. A _Handler_ is a mapping between a specific set of DWN message properties and an HTTP API endpoint exposed by the downstream service. e.g. "When a DWN message is received that matches the filter: `{ "method": "CollectionsWrite", "schema": "https://tbdex.io/schemas/Ask", "protocol": "tbdex" }` send it to `POST https://somehost.com/some-endpoint`". Handlers are configured via a config file that is loaded by the relay service on startup.
+
+```mermaid
+sequenceDiagram
+  autonumber
+  participant C as Client
+  participant R as DWN Relay
+  participant S as Downstream Service
+  
+  C->>R: DWeb Message
+  R->>R: Integrity Checks
+  R->>R: Lookup registered handlers
+  R->>S: POST /${registered-handler}
+  S->>S: Handle Request
+  S->>R: HTTP Response
+  R->>R: DWM'ify Response
+  R->>C: DWMified response
+  
+```
+
+# Handler Registration 
+
+## Config file
+
+⚠  **TODO**: Include every option
+```json
+{
+  "handlers": [
+    {
+      "filter": {
+        "method": "CollectionsWrite",
+        "schema": "https://ssi-schemas.org/CredentialApplication",
+        "protocol": ""
+      },
+      "endpoint": {
+        "method": "POST",
+        "url": "https://somehost/some-endpoint",
+        "requestOptions": {},
+        "responseMapping": {
+          "201": {
+            "method": "CollectionsWrite",
+            "schema": "https://ssi-schemas.org/CredentialApplication"
+          }
+        }
+      }
+    }
+  ]
+}
+```
+
+# DWN Relay Message Handling
+On ingress (step 1 in the sequence diagram), the DWN Relay Service will:
+  - leverage the DWN SDK to process and store the message
+    - ⚠  **TODO**: Decide on whether storage can be toggled off. Instinctively I think it makes sense to be able to toggle it off but that does have considerable implications. The most notable implication is that `protocol` cannot be adhered to without storage
+  - base64url decode / decrypt `encodedData` within the message
+    - ⚠  **TODO**: include details regarding which keys are being used to decode/decrypt the message
+    - ⚠  **TODO**: include details regarding what happens if decoding/decrypting fails
+  - send an HTTP request to the appropriate handler. Details about the request body are described [here](#dwn-relay---downstream-service-request-payload)
+
+## Example Message
+An example of a DWeb Message that a DWN Relay would receive from a client
```

Comment:
need the Client -> DWN Relay header here

Reply:
doneski